ibis/expr/builders.py

```diff
@@ -176,6 +176,8 @@ def _maybe_cast_boundary(self, boundary, dtype):
 
     def _maybe_cast_boundaries(self, start, end):
         if start and end:
+            if start.dtype.is_same_base_type(end.dtype):
```

Comment:
Could you please cover this case in `ibis/expr/tests` so that `pytest -m core` can capture if there is a regression?